circle/src/util.rs

```diff
@@ -0,0 +1,168 @@
+    if log_n == 0 {
+        return vec![F::one()];
+    }
+    let mut basis = vec![F::one()];
+    basis.reserve(1 << log_n);
+    let mut cur = point.real();
+    for _ in 0..(log_n - 1) {
+        for i in 0..basis.len() {
+            basis.push(basis[i] * cur);
+        }
+        cur = F::two() * cur.square() - F::one();
+    }
+    reverse_slice_index_bits(&mut basis);
+    for i in 0..basis.len() {
+        basis.push(basis[i] * point.imag());
+    }
+    for (i, val) in basis.iter_mut().enumerate() {
+        let num_adjacent_ones = (i & (i >> 1)).count_ones();
+        if num_adjacent_ones % 2 == 1 {
+            *val = -*val;
+        }
+    }
+    basis
+}
+
+#[cfg(test)]
+pub(crate) fn eval_circle_polys<F: ComplexExtendable>(
+    coeffs: &RowMajorMatrix<F>,
+    point: Complex<F>,
+) -> Vec<F> {
+    let log_n = log2_strict_usize(coeffs.height());
+    let mut accs = vec![F::zero(); coeffs.width()];
+    for (row, basis) in coeffs.rows().zip(circle_basis(point, log_n)) {
+        for (acc, coeff) in izip!(&mut accs, row) {
+            *acc += *coeff * basis;
+        }
+    }
+    accs
+}
+
+// Page 4 of the paper:
+// (x, y) = ((1-t^2)/(1+t^2), 2t/(1+t^2))
+// t = ±i goes to (-1, 0)
+pub(crate) fn univariate_to_point<F: Field>(t: F) -> Complex<F> {
+    let t2 = t.square();
+    if let Some(inv_denom) = (F::one() + t2).try_inverse() {
+        Complex::new((F::one() - t2) * inv_denom, (F::two() * t) * inv_denom)
+    } else {
+        Complex::new_real(F::neg_one())
+    }
+}
+
+// t = y / (x + 1)
+// If F has i, this should return that instead, but we don't have access..
+pub(crate) fn point_to_univariate<F: Field>(p: Complex<F>) -> Option<F> {
+    Some(p.imag() * (p.real() + F::one()).try_inverse()?)
+}
+
+// Page 5
+// same as above, this *could* handle point at infinity if we had Field::try_sqrt
+#[allow(unused)]
+pub(crate) fn rotate_univariate<F: Field, EF: ExtensionField<F>>(t1: EF, t2: F) -> Option<EF> {
+    Some((t1 + t2) * (EF::one() - t1 * t2).try_inverse()?)
+}
+
+// the zero poly for cfft domain sized log_n, evaluated at p_x
+pub(crate) fn v_n<F: Field>(mut p_x: F, log_n: usize) -> F {
+    for _ in 0..(log_n - 1) {
+        p_x = F::two() * p_x.square() - F::one();
+    }
+    p_x
+}
+
+fn v_n_prime<F: Field>(p_x: F, log_n: usize) -> F {
+    F::two().exp_u64((2 * (log_n - 1)) as u64) * (1..log_n).map(|i| v_n(p_x, i)).product()
+}
+
+// Simple zero at (1,0), simple pole at (-1,0)
+// panics if called with x = -1
+pub(crate) fn v_0<F: Field>(p: Complex<F>) -> F {
+    p.imag() / (p.real() + F::one())
+}
+
+pub(crate) fn s_p_at_p<F: Field>(p_x: F, p_y: F, log_n: usize) -> F {
+    -F::two() * v_n_prime(p_x, log_n) * p_y
+}
+
+// tranposed matrix-vector product: Mᵀv
+// TODO: pack (currently ~100ms of m31_keccak)
+#[instrument(skip_all, fields(dims = %m.dimensions()))]
+pub(crate) fn gemv_tr<'a, F: Field, EF: ExtensionField<F>>(
```

Comment:
We can parallelize this function across the outer loop, and have a feature flag parallel for it